### PR TITLE
feat(ecr): adding ECR pull through cache rule

### LINKED ISSUE #20464: (ecr): PullThroughCacheRule
### Describe the feature

Implementation of AWS::ECR::PullThroughCacheRule https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecr-pullthroughcacherule.html

### Use Case

Looks like this is missing

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.24.1

### Environment details (OS name and version, etc.)

MacOS 12.3.1, Python 3.9

### PR DESCRIPTION
**Pull request by daschaa**
Fixes #20464 
 
With this PR I tried to implement the pull through cache rules from ECR. (See here: [https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-through-cache.html](https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-through-cache.html))

I was not sure if this needs a L2 Construct but I thought it would be good, because it is then shown in the CDK reference documentation. L1 Constructs are, in my experience, a bit more hidden.
I'm still not sure if we should expose it to the user, because the list of things to consider when creating such a rule is quite long. I've added a hint in the docs and decided against copying all bullet points to the docs. 

Some open questions:
- Should there be a method which automatically grants the `ecr:CreatePullThroughCacheRule` (and others) on a `Grantee` ? 
- Should there be more documentation? I've copied the text from the original AWS docs and I'm not sure if this is enough.

I would appreciate your feedback! ❤️ 

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by corymhall — DISMISSED** [review]
I haven't looked into pull through cache in detail, but I'm thinking we may want to think about having this extend `IRepository`. The way I understand it it, each image you pull through a cache rule gets it's own private repository.

I would suggest as next steps to iterate on the README so we can settle on the user experience that we want. Some things to think about:

1. If I create a pull through cache, how do I reference an image in an ECS application (for example).
2. If I want to restrict access to the private repository through a repo policy, how do I do that?
3. Is there a better way of modeling the input properties? From the console you can only select ecr-public or quay and each have a specific Url value. Maybe something like `upstream: Upstream.ecrPublic('my-ecr')` etc.

**Comment by daschaa** [comment]
@corymhall 
I added a section in the readme for the usage of the repository created by the pull through cache rule. I took the example from this re:Invent talk actually: [AWS re:Invent 2021 - Up-level your container image security with the latest from Amazon ECR](https://www.youtube.com/watch?v=gvzyF0WxQao&t=2083s).

For the access restriction I thought it would be good to have a function where I can put in some principals and repositories and the registry policy is generated out of that. However I'm not sure if that is too 'restrictive'. And I'm not sure if `IPrincipal` is the best interface there, because it **has to** be a user/root principal, otherwise the registry policy won't deploy. (I have learned it the hard way while executing the integration tests).

For the upstream urls I decided to use an enum, because it felt naturally to me. And it divides the 'hard-coded' upstream url and the customizable repository prefix in two parts :)

Next step would be to add more tests and iterate on the `restrictAcess` functionality, because right now it does not feel very CDK-ish. I thought about using `grant..` but I don't know if this is the cleanest way either.

If you have any feedback on the current changes I would be very happy :) Thank you as always for the valuable feedback!

File: packages/@aws-cdk/aws-ecr/test/pull-through-cache-rule.test.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                ],
+              },
+            ],
+          },
+        ],
+        Version: '2012-10-17',
+      },
+    });
+  });
+});
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ecr/test/pull-through-cache-rule.test.ts:168]:
```suggestion
});

```

File: packages/@aws-cdk/aws-ecr/test/integ.pull-through-cache-rule.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+/// !cdk-integ aws-ecr-pull-through-cache-rule
+import * as cdk from '@aws-cdk/core';
+import * as integ from '@aws-cdk/integ-tests';
+import * as ecr from '../lib';
+import { UpstreamRegistry } from '../lib';
+
+const app = new cdk.App();
+
+const stack = new cdk.Stack(app, 'aws-ecr-pull-through-cache-rule');
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-ecr/test/integ.pull-through-cache-rule.ts:9]:
```suggestion
const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-ecr-pull-through-cache-rule');
```

**Review by comcalvi — CHANGES_REQUESTED** [review]

**Pull request closed without merging**